### PR TITLE
Structured name mangling

### PR DESCRIPTION
**Pull request by densh**
Previously we used to mangle names early in the
pipeline using a lossy conversion to strings that
couldn't be reversed.

This pr introduces a few things to improve it:

1. Introduce new name mangling scheme with a small spec.

2. Global names now incorporate signatures, an additional
   data structure that lets one programmatically see additional
   information about the name.

3. Adds a compelete unmangler to make sure that we can
   go back to the unmangled representation from the mangled
   one.

### CODE REVIEW

**Comment by densh** [comment]
@jonas It would be nice if you could have a pass over this one.

**Comment by jonas** [comment]
I've started reviewing. Give me until Wednesday morning Lausanne time to get through it.

**Comment by densh** [comment]
@jonas Thanks! It's critical to get this right, we can't change the name mangling scheme very often. It's better to take our time with this one. 

Also, just added one more missing piece: support for nullability information. We don't use it yet, but it's necessary for Interflow (#1339) and it does affect name mangling as well.

File: javalib/src/main/scala/java/lang/StackTraceElement.scala [hunk 1]
```diff
@@ -37,21 +38,71 @@ final class StackTraceElement(val getClassName: String,
...
+    }
+
+    def readNumber(): Int = {
+      val start = pos
+      var char  = chars(pos)
+      while ('0' <= char && char <= '9') {
+        pos += 1
+        char = chars(pos)
       }
+      java.lang.Integer.parseInt(sym.substring(start, pos))
```
[the remark this answers is not in the document]
Comment (densh) [reply javalib/src/main/scala/java/lang/StackTraceElement.scala]: Agreed!

File: docs/contrib/mangling.rst [hunk 2]
```diff
@@ -0,0 +1,50 @@
+Name mangling
+-------------
+
+Scala Native toolchain mangles names for all definitions except
+the ones which have been explicitly exported to C using
+``extern``. Mangling scheme is defined through a simple grammar
+thar uses a notation inspired by
```
[the remark this answers is not in the document]
Comment (densh) [reply docs/contrib/mangling.rst]: Good catch!

File: docs/contrib/mangling.rst [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        <integer-type-name>            // signed integer type-name
+        z                              // scala.Boolean
+        c                              // scala.Char
+        f                              // scala.Float
+        d                              // scala.Double
+        u                              // scala.Unit
+        n                              // scala.Nothing
+        L <nullable-type-name>         // nullable type-name
+        A <type-name> _                // nonnull array type-name
+        X <name>                       // nonnull exact class type-name
```
[the remark this answers is not in the document]
Comment (densh) [reply docs/contrib/mangling.rst]: Given a pseudo-Scala syntax with exact types:
```
class A { def foo = 'A' }
class B extends A { override def foo = 'B' }

val x: exact A = new A
x.foo // can *only* call A.foo

val y: A = x
x.foo // can call either A.foo or B.foo
```
From the Scala point of view all reference types are non-exact and nullable. Interflow will infer exactness and nullability. Exactness is primiraly useful for devirtualization. Nullability for partial evaluation of is/as instance of.

File: nir/src/main/scala/scala/scalanative/nir/Defns.scala [hunk 4]
```diff
@@ -20,8 +20,6 @@ object Defn {
                           ty: Type,
                           insts: Seq[Inst])
       extends Defn
-  final case class Struct(attrs: Attrs, name: Global, tys: Seq[Type])
-      extends Defn
```
[the remark this answers is not in the document]
Comment (densh) [reply nir/src/main/scala/scala/scalanative/nir/Defns.scala]: Yes, but they are going to come back in one of the next PRs before 0.4.0 is released.

File: docs/contrib/mangling.rst [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+        <name>                         // nonnull class type-name
```
[the remark this answers is not in the document]
Comment (densh) [reply docs/contrib/mangling.rst]: 👍

File: tools/src/test/scala/scala/scalanative/nir/TypeManglingSuite.scala [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    Type.Function(Seq(Type.Float, Type.Int), Type.Int),
+    Type.Nothing,
+    Type.Unit,
+    Type.Array(Rt.Object, nullable = false),
+    Type.Array(Rt.Object, nullable = true),
+    Type.Ref(Rt.Object.name, exact = true, nullable = true),
+    Type.Ref(Rt.Object.name, exact = true, nullable = false),
+    Type.Ref(Rt.Object.name, exact = false, nullable = true),
+    Type.Ref(Rt.Object.name, exact = false, nullable = false)
+  ).foreach { ty =>
```
[the remark this answers is not in the document]
Comment (densh) [reply tools/src/test/scala/scala/scalanative/nir/TypeManglingSuite.scala]: 👍

File: docs/contrib/mangling.rst [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+        G <name>                       // generated name
+        K <sig-name> <type-name>+ E    // duplicate name
+
+    <type-name> ::=
+        v                              // c void
+        g                              // c vararg
+        R _                            // c pointer type-name
+        R <type-name>+ E               // c function type-name
+        S <type-name>+ E               // c anonymous struct type-name
+        A <type-name> <number> _       // c array type-name
```
[the remark this answers is not in the document]
Comment (densh) [reply docs/contrib/mangling.rst]: Actually, we should rather fix the code. We've used `E` only after the variable-length number of elements as end terminator so far. It should probably stay this way.

**Comment by densh** [comment]
Ready for the second pass.

**Comment by densh** [comment]
@jonas Any more comments on this one?

File: tools/src/test/scala/scala/scalanative/linker/ReachibilitySuite.scala [hunk 8]
```diff
@@ -5,13 +5,16 @@ import org.scalatest._
 import java.io.File
 import java.nio.file.{Files, Path, Paths}
 import scalanative.util.Scope
-import scalanative.nir.Global
+import scalanative.nir.{Sig, Global}
 import scalanative.build.ScalaNative
 
 trait ReachabilitySuite extends FunSuite {
```
[the remark this answers is not in the document]
Comment (densh) [reply tools/src/test/scala/scala/scalanative/linker/ReachibilitySuite.scala]: 👍

**Comment by densh** [comment]
@jonas Itanium ABI was mostly an inspiration for notation and organization of the spec. If it's the spec is clear by itself -- that's great actually.

**Comment by densh** [comment]
@jonas Thanks for the review! 👍

**Pull request merged**